docs/tutorials/05_kv-scheduler.md

```diff
@@ -0,0 +1,303 @@
+
+The vpp-agent uses VPP API in form of binary API calls. Each VPP binary API call is designed to create a configuration item in the VPP or add or modify various parameters. In practice, these actions can be dependent on each other. The IP address can be assigned to an interface only if the interface is already present in the VPP. Another example is a VPP forwarding FIB entry, which can be added only if required interface and bridge domain exists and the interface is also assigned in the bridge domain, creating a complex dependency tree. In general, it is true that:
+
+1. to configure all the proto-modelled data provided by the northbound, usually more than one binary API call is required
+2. to configure parameters, a parent item is required to be present
+2. some configuration items themselves are dependent on other and cannot be configured earlier
+
+ It means that binary API calls have to be called in an exact order. The VPP agent uses a KV scheduler component to ensure it with system of caching and configuration dependency handling. Every plugin configuring something dependent on other plugin's configs in certain way can be added to the KV scheduler and profit from its advantages. 
+ 
+ As a first step, we will use the special [proto model][1]. The model defines two simple messages - an `Interface` and a `Route` requiring some interface. The model demonstrates simple dependency between configuration items (since we need an interface to configure the route).  
+ 
+ In order to register our hello world plugin to the scheduler and work with the new model, we need a two new components - a **descriptor** and an **adapter** for every proto-defined type.
+ 
+ #### 1. Adapters
+ 
+ Let's start with adapters. The purpose of the adapter is to define conversion methods between our proto-defined type and bare `proto.Message`. Since this code fulfils the definition of a boilerplate, we will generate it. The generator is called `descriptor-adapter` and can be found in the [inside the KVScheduler plugin][2]. Build the binary file from the go files inside, and use it to generate two adapters for the `Interface` and `Route` proto messages:
+ 
+```
+descriptor-adapter --descriptor-name Interface --value-type *model.Interface --import "github.com/ligato/vpp-agent/examples/tutorials/05_kv-scheduler/model" --output-dir "descriptor"
+descriptor-adapter --descriptor-name Route --value-type *model.Route --import "github.com/ligato/vpp-agent/examples/tutorials/05_kv-scheduler/model" --output-dir "descriptor"
+```
+
+It is a good practice to add those commands to the plugin file with the `//go:generate` directives. Adapters were automatically created in the `adapter` directory within plugin folder.
+
+#### 2. Descriptor without dependency
+
+Another step is to define descriptors. We start with the interface descriptor, since there is no dependency on other messages. Let's create the `descriptors.go` so the code is outside of `main.go`, and define following code:
+
```

Comment:
maybe just mention that the preferred way is to have separate .go file for every descriptor.

Reply:
added